src/lib/useEmojis.ts

```diff
@@ -9,20 +9,29 @@ import type {
 
 function reducer(state: EmojiiState, action: EmojiAction) {
   const { emoji } = action;
-  const emojiFromState = state.find((em) => em.emoji === emoji.emoji) || emoji;
+  const stateEmoji = state.find((em) => em.emoji === emoji.emoji);
+  const emojiFromState = stateEmoji || emoji;
   emojiFromState.counter = emojiFromState.counter || 0;
   switch (action.type) {
     case 'i':
       emojiFromState.counter++;
       break;
     case 'd':
       emojiFromState.counter--;
-      if (emojiFromState.counter < 0) emojiFromState.counter = 0;
+      if (emojiFromState.counter <= 0) emojiFromState.counter = 0;
   }
-  return [
-    ...state.filter((rea) => rea.emoji !== emojiFromState.emoji),
-    ...(emojiFromState.counter > 0 ? [emojiFromState] : []),
-  ];
+  if (!stateEmoji) {
+    state.push(emojiFromState);
+  }
+
+  return state
+    .map((rea) => {
+      if (rea.emoji === emojiFromState.emoji) {
+        return emojiFromState.counter === 0 ? null : emojiFromState;
+      }
+      return rea;
```

Comment:
I'll try to create some tests for this lines, hopefully it won't be too complicated

Reply:
great!